Add a latest command that shows the latest released version

mattwynne opened the pull request:
I wanted a way to get the latest released version number out of my CHANGELOG.md

I hope this is useful!

I've test-driven the changes, so hopefully it should all be good for any refactoring.

I did consider adding this as a default to the `show` command if you didn't specify a version to show, but that command doesn't actually show the version number, which is what I need.

This does allow you to do ``changelog show `changelog latest` `` to show the details of the latest release.

Note that this is my first Go code, so:

a) please be gentle with your feedback
b) thanks for writing such an easy-to-navigate codebase that I could learn from and so easily contribute to!

mattwynne committed:
Add a latest command that shows the latest released version

mattwynne committed:
Clarify help text

mattwynne committed:
Add another test case and refactor

mattwynne commented on cmd/latest.go:
```diff
@@ -0,0 +1,43 @@
+	}
+}
+
+func releasedVersions(changelog *chg.Changelog) (result []chg.Version) {
```
Potentially this could be moved onto the ChangeLog model?

rcmachado replied:
Agreed - cmd files are a little bit bloated. If we can keep new logic out of it would help

mattwynne replied:
OK that's done in 50c3258

mattwynne committed:
Update README

mattwynne committed:
Add changelog entry for #77

rcmachado commented:
@mattwynne Thanks for your contribution - really good! Just one minor comment - if you could move that logic out of the command it would be great and then I think we are ready to merge

Edit: just authorized the checks and it seems some of them failed - do you mind checking, please?

Commit:
Move ReleasedVersions onto ChangeLog model

mattwynne commented:
Thanks for taking a look @rcmachado and sorry for the lag - my GitHub notifications were going into a folder I don't look in very often.

I'm a bit confused by the check failures. The `lint` task shows `golangci-lint found no issues` so I'm not sure why that counts a failure.

The goreleaser error looks like maybe their manifest format has changed between versions of goreleaser? The error says:

    line 13: field builds not found in type config.Docker
    line 15: field binaries not found in type config.Docker

mattwynne commented:
@rcmachado I'd appreciate some help with the linter / goreleaser warnings if you can spare the time. I'm a bit stuck understanding their feedback.

aslakhellesoy commented:
I think it would be more unix'y (i.e. versatile) if this were refactored to a `list` command that lists *all* releases. That would be a command that can be used for a lot more. The output could be piped to `grep`, `head`, `tail`, `cut` etc to extract a particular release.

Suggested output from `changelog list` ([TSV](https://en.wikipedia.org/wiki/Tab-separated_values)):

```
Unreleased\t\thttp://example.com/1.0.0..HEAD\t
1.0.0\t2018-04-23\thttp://example.com/abcdef..1.0.0\ttrue
```

The column headers would not be printed (just like `ls -al` doesn't print the column headers), but they would be:

```
Name\tDate\nLink\nYanked
```

aslakhellesoy committed:
Ignore .idea folder

aslakhellesoy committed:
Make Changelog JSONable

aslakhellesoy committed:
Add --json flag to show command

aslakhellesoy committed:
Add --json to fmt command

aslakhellesoy commented:
I thought about this some more, and implemented #78 which adds a `--json` option to `fmt` and `show`. That makes it easy to get the latest version (and more) by piping to `jq`.

WDYT?

aslakhellesoy committed:
Output change type as string instead of int

mattwynne commented:
I think a `list` that just listed all the versions could be useful - you could then use `show` to look at a particular one.

The JSON output seems like a great idea to me, very versatile! Can we have both!?

aslakhellesoy commented:
@mattwynne we don't need a new `list` command. The `fmt --json` command lists the entire changelog as JSON. See the example in #78

aslakhellesoy committed:
Update docs

aslakhellesoy committed:
Update changelog

aslakhellesoy committed:
Merge pull request #1 from cucumber/json-output

Json output

aslakhellesoy committed:
Update badges and mention this is a fork

aslakhellesoy committed:
Rename module and update dependencies

aslakhellesoy committed:
Add not about the fork

aslakhellesoy committed:
Disable goimports since I don't know how to make it pass

rcmachado commented:
@mattwyne sorry for the delay - time has been a scarce resource lately.

About CI process failing, couldn't take a look yet but could be something related to tools been updated and rules changed? I don't build this project since some time

aslakhellesoy committed:
Add --tag-pattern option to the release command

aslakhellesoy committed:
Rename --tag-pattern to --tag-format

aslakhellesoy committed:
Update docs

mattwynne committed:
Merge pull request #2 from cucumber/release-tag-pattern

Add --tag-format option to the release command

mattwynne committed:
Remove deprecated fields from goreleaser config

mattwynne committed:
Merge pull request #3 from cucumber/fix-goreleaser

Remove deprecated fields from goreleaser config

mattwynne committed:
Release on changes to main branch

Not sure why we'd want to release a PR

mattwynne committed:
Update CHANGELOG.md

mattwynne committed:
Release 0.8.0

mattwynne committed:
Fix `show` command

Fixes #4

mattwynne committed:
Update changelog for #4

mattwynne committed:
Release 0.9.0

mattwynne committed:
Add an ARM64 build too

mattwynne committed:
Update changelog

mattwynne committed:
Configure release action with correct environment / secrets

mattwynne committed:
Release 0.9.2

mattwynne committed:
Update CHANGELOG

mattwynne committed:
Split goreleaser action into two jobs

The `build` job will run on a PR or a push to `main`. The `release` job
just runs on a tag.

mattwynne committed:
Merge branch 'main' into show-latest-version

mattwynne committed:
Fix ref to parser module

mattwynne committed:
Reformat for linter

mattwynne committed:
Release 0.10.0

tina-junold commented:
Any news on this? I would like to see this.

mattwynne commented:
Hey @tburschka. We have this implemented in our fork: https://github.com/cucumber/changelog
